### ISSUE TITLE
Handwritten Digit Recognition with AI- SHEELA with DeepSeek AI

### ISSUE THREAD

**Issue by minimindsai** [minimindsai/Deepseek-MinimindsAI#1]
Project Overview:
This project will use the popular MNIST dataset to train an AI model to recognize handwritten digits (0-9). It will involve data preprocessing, model training using TensorFlow/Keras, evaluation, and deployment via a simple web interface.

Project Structure:
sql
Copy
Edit
AI-Handwritten-Digit-Recognition/
│-- README.md
│-- requirements.txt
│-- dataset/
│   └── mnist_data.npz
│-- notebooks/
│   └── EDA.ipynb
│-- src/
│   ├── data_preprocessing.py
│   ├── model_training.py
│   ├── evaluation.py
│   ├── predict.py
│   └── utils.py
│-- web_app/
│   ├── app.py
│   ├── templates/
│   │   ├── index.html
│   ├── static/
│   │   ├── styles.css
│-- models/
│   ├── trained_model.h5
│-- tests/
│   ├── test_model.py
└── .gitignore
Step-by-Step Implementation:
1. Data Preprocessing (src/data_preprocessing.py)
python
Copy
Edit
import tensorflow as tf
from tensorflow.keras.datasets import mnist
import numpy as np

def load_data():
    (x_train, y_train), (x_test, y_test) = mnist.load_data()
    x_train, x_test = x_train / 255.0, x_test / 255.0  # Normalize data
    x_train = x_train.reshape(-1, 28, 28, 1)
    x_test = x_test.reshape(-1, 28, 28, 1)
    return (x_train, y_train), (x_test, y_test)
2. Model Training (src/model_training.py)
python
Copy
Edit
import tensorflow as tf
from tensorflow.keras import layers, models
from data_preprocessing import load_data

(x_train, y_train), (x_test, y_test) = load_data()

model = models.Sequential([
    layers.Conv2D(32, (3,3), activation='relu', input_shape=(28,28,1)),
    layers.MaxPooling2D((2,2)),
    layers.Conv2D(64, (3,3), activation='relu'),
    layers.MaxPooling2D((2,2)),
    layers.Flatten(),
    layers.Dense(64, activation='relu'),
    layers.Dense(10, activation='softmax')
])

model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

model.fit(x_train, y_train, epochs=5, validation_data=(x_test, y_test))
model.save('../models/trained_model.h5')
3. Model Evaluation (src/evaluation.py)
python
Copy
Edit
from tensorflow.keras.models import load_model
from data_preprocessing import load_data

model = load_model('../models/trained_model.h5')
(_, _), (x_test, y_test) = load_data()

test_loss, test_acc = model.evaluate(x_test, y_test, verbose=2)
print(f"Test accuracy: {test_acc:.2f}")
4. Model Prediction (src/predict.py)
python
Copy
Edit
import numpy as np
from tensorflow.keras.models import load_model
import cv2

model = load_model('../models/trained_model.h5')

def predict_digit(img_path):
    img = cv2.imread(img_path, cv2.IMREAD_GRAYSCALE)
    img = cv2.resize(img, (28, 28)) / 255.0
    img = img.reshape(1, 28, 28, 1)
    prediction = model.predict(img)
    return np.argmax(prediction)

print("Predicted digit:", predict_digit('sample_digit.png'))
5. Web App (web_app/app.py - Flask Framework)
python
Copy
Edit
from flask import Flask, render_template, request
import numpy as np
import cv2
from src.predict import predict_digit

app = Flask(__name__)

@app.route('/', methods=['GET', 'POST'])
def index():
    if request.method == 'POST':
        file = request.files['file']
        file_path = f"./static/{file.filename}"
        file.save(file_path)
        result = predict_digit(file_path)
        return render_template('index.html', prediction=result, image=file.filename)
    return render_template('index.html', prediction=None)

if __name__ == '__main__':
    app.run(debug=True)
6. HTML Interface (web_app/templates/index.html)
html
Copy
Edit
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Digit Recognizer</title>
</head>
<body>
    <h1>Upload Handwritten Digit</h1>
    <form action="/" method="post" enctype="multipart/form-data">
        <input type="file" name="file" required>
        <button type="submit">Predict</button>
    </form>
    {% if prediction is not none %}
        <h2>Predicted Digit: {{ prediction }}</h2>
        <img src="{{ url_for('static', filename=image) }}" width="200">
    {% endif %}
</body>
</html>
7. Requirements File (requirements.txt)
Copy
Edit
tensorflow
numpy
opencv-python
flask
How to Run the Project:
Clone the repository:

bash
Copy
Edit
git clone https://github.com/yourusername/AI-Handwritten-Digit-Recognition.git
cd AI-Handwritten-Digit-Recognition
Install dependencies:

bash
Copy
Edit
pip install -r requirements.txt
Train the model:

bash
Copy
Edit
python src/model_training.py
Run the web application:

bash
Copy
Edit
python web_app/app.py
Open in the browser:
Navigate to http://127.0.0.1:5000/ and upload an image.

Simplified Explanation with Images
1. Data Preprocessing:

2. Model Architecture:

3. Prediction Process:

Educational Benefits of the Project:
Understanding deep learning concepts through practical implementation.
Learning about data preprocessing and normalization.
Model evaluation and optimization techniques.
Deployment of AI models using Flask for web applications.
Building an end-to-end AI pipeline.

**Comment by minimindsai** [comment]
import matplotlib.pyplot as plt
from matplotlib.patches import FancyBboxPatch
from matplotlib.lines import Line2D
from matplotlib.patheffects import withStroke

def create_sheela_ai_chart():
    # Create figure and axis
    fig, ax = plt.subplots(figsize=(12, 8))
    ax.axis('off')

    # Define chart components
    components = {
        "SHEELA AI Core": (0.5, 0.85),
        "Real-Time Trend Analysis": (0.2, 0.7),
        "Mood and Tone Adjustment": (0.8, 0.7),
        "Dynamic Content Creation": (0.2, 0.5),
        "Advanced Media Insights": (0.5, 0.5),
        "Twitter Engagement": (0.8, 0.5),
        "AI Coin Rewards": (0.5, 0.3),
        "User Analytics Dashboard": (0.5, 0.1)
    }

    connections = [
        ("SHEELA AI Core", "Real-Time Trend Analysis"),
        ("SHEELA AI Core", "Mood and Tone Adjustment"),
        ("Real-Time Trend Analysis", "Dynamic Content Creation"),
        ("Mood and Tone Adjustment", "Twitter Engagement"),
        ("Dynamic Content Creation", "Advanced Media Insights"),
        ("Twitter Engagement", "Advanced Media Insights"),
        ("Advanced Media Insights", "AI Coin Rewards"),
        ("AI Coin Rewards", "User Analytics Dashboard")
    ]

    # Colors and aesthetics
    box_color = "#8ecae6"
    highlight_color = "#ffb703"
    text_color = "white"
    line_color = "#219ebc"
    gradient_effect = withStroke(linewidth=3, foreground="#023047")

    # Draw components with enhanced visuals
    for label, (x, y) in components.items():
        is_core = label == "SHEELA AI Core"
        box = FancyBboxPatch(
            (x - 0.12, y - 0.05), 0.24, 0.1, 
            boxstyle="round,pad=0.2", color=highlight_color if is_core else box_color, 
            ec=line_color, lw=2, path_effects=[gradient_effect]
        )
        ax.add_patch(box)
        ax.text(x, y, label, ha="center", va="center", fontsize=12 if is_core else 10, 
                color=text_color, weight="bold")

    # Draw connections with smooth and styled lines
    for start, end in connections:
        start_x, start_y = components[start]
        end_x, end_y = components[end]
        ax.add_line(Line2D(
            [start_x, end_x], [start_y, end_y], 
            color=line_color, lw=2.5, alpha=0.9, linestyle="-", 
            path_effects=[gradient_effect]
        ))

    # Add title and subtitle
    plt.title(
        "SHEELA AI: Adaptive Twitter Management Ecosystem",
        fontsize=16, color="#023047", weight="bold", pad=20
    )
    ax.text(
        0.5, 0.9, "An Integrated Approach to AI-Powered Social Media Engagement", 
        fontsize=12, color="#023047", ha="center", style="italic"
    )

    # Annotations for enhanced understanding
    annotations = {
        "SHEELA AI Core": "Central hub for AI decision-making",
        "Real-Time Trend Analysis": "Monitors and analyzes global news",
        "Mood and Tone Adjustment": "Adapts tone to match current events",
        "Dynamic Content Creation": "Generates relevant and engaging posts",
        "Advanced Media Insights": "Provides deeper analytics and trends",
        "Twitter Engagement": "Optimizes interactions and reach",
        "AI Coin Rewards": "Incentivizes contributions and activity",
        "User Analytics Dashboard": "Summarizes performance and metrics"
    }

    for label, description in annotations.items():
        x, y = components[label]
        ax.text(x, y - 0.08, description, fontsize=8, ha="center", color="#023047")

    # Display the chart
    plt.tight_layout()
    plt.show()

# Call the function to display the chart
create_sheela_ai_chart()
